js/ui/windowManager.js

```diff
@@ -2166,3 +1888,368 @@ const WindowManager = new Lang.Class({
+        * window's front-face is completely invisible to the user
+        * (this cannot be done just by looking at the angle of the
+        * window because the same angle may show a different visible
+        * face depending on its position in the view frustum).
+        *
+        * What we do here is enable backface culling and rotate both
+        * windows by 180degrees. The effect of this is that the front
+        * and back window will be at opposite rotations at each point
+        * in time and so the exact point at which the first window
+        * becomes invisible is the same point at which the second
+        * window becomes visible. Because no back faces are drawn
+        * there are no visible artifacts in the animation */
+        src.set_cull_back_face(true);
+        dst.set_cull_back_face(true);
+
+        src.show();
+        dst.show();
+        dst.opacity = 0;
+
+        let dst_geometry = src.get_meta_window().get_frame_rect();
+        dst.get_meta_window().move_resize_frame(false,
+                                                dst_geometry.x,
+                                                dst_geometry.y,
+                                                dst_geometry.width,
+                                                dst_geometry.height);
+
+
+        /* Tween both windows in a rotation animation at the same time
+        * with backface culling enabled on both. This will allow for
+        * a smooth transition. */
+        Tweener.addTween(src, {
+                        rotation_angle_y: direction == Gtk.DirectionType.RIGHT ? 180 : -180,
+                        time: WINDOW_ANIMATION_TIME * 4,
+                        transition: 'easeOutQuad',
+                        onComplete: Lang.bind(this, function() {
+                            this.rotateOutCompleted(src);
+                        })
+        });
+        Tweener.addTween(dst, {
+                        rotation_angle_y: 0,
+                        time: WINDOW_ANIMATION_TIME * 4,
+                        transition: 'easeOutQuad',
+                        onComplete: Lang.bind(this, function() {
+                            this.rotateInCompleted(dst);
+                        })
+        });
+
+        /* Gently fade the window in, this will paper over
+         * any artifacts from shadows and the like */
+        Tweener.addTween(dst, {
+                    opacity: 255,
+                    time: WINDOW_ANIMATION_TIME * 4,
+                    transition: 'linear'
+        });
+    },
+
+    _removeEffect : function(list, actor) {
+        let idx = list.indexOf(actor);
+        if (idx != -1) {
+            list.splice(idx, 1);
+            return true;
+        }
+        return false;
+    },
+
+    _removeSession : function(session) {
+        let idx = this._sessions.indexOf(session);
+        if (idx != -1) {
+            this._sessions.splice(idx, 1);
+            return true;
+        }
+        return false;
```

Comment:
Can this call out to `_removeEffect` or be refactored so that both call the same function?